### ISSUE TITLE
Sik-Ho Tang | Review: SqueezeNet (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#99]
Sik-Ho Tang. [Review: SqueezeNet (Image Classification)](https://towardsdatascience.com/review-squeezenet-image-classification-e7414825581a).

**Comment by NorbertZheng** [comment]
## Overview
In this story, SqueezeNet, by DeepScale, UC Berkeley and Stanford University, is reviewed. With equivalent accuracy, smaller CNN architectures offer at least three advantages:
 - Smaller Convolutional Neural Networks (CNNs) require less communication across servers during distributed training.
 - Smaller CNNs require less bandwidth to export a new model from the cloud to an autonomous car.
 - Smaller CNNs are more feasible to deploy on FPGAs and other hardware with limited memory.

This is a technical report on arXiv in 2016 with over 1100 citations.

**Comment by NorbertZheng** [comment]
## Architectural Design Strategies
### Strategy 1. Replace 3×3 filters with 1×1 filters
Given a budget of a certain number of convolution filters, we can choose to make the majority of these filters 1×1, since a 1×1 filter has 9× fewer parameters than a 3×3 filter.
### Strategy 2. Decrease the number of input channels to 3×3 filters
Consider a convolution layer that is comprised entirely of 3×3 filters. The total quantity of parameters in this layer is:
 - (number of input channels) × (number of filters) × (3×3)

We can decrease the number of input channels to 3×3 filters using squeeze layers, mentioned in the next section.
### Strategy 3. Downsample late in the network so that convolution layers have large activation maps
The intuition is that large activation maps (due to delayed downsampling) can lead to higher classification accuracy.

### Summary
 - Strategies 1 and 2 are about judiciously decreasing the quantity of parameters in a CNN while attempting to preserve accuracy.
 - **Strategy 3 is about maximizing accuracy on a limited budget of parameters.**

**Comment by NorbertZheng** [comment]
## Fire Module
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/db7ceaa3-32e9-4dc6-8b50-1746be1d8f5c)
Fire Module with hyperparameters: s1x1 = 3, e1x1 = 4, and e3x3 = 4.

 - A Fire module is comprised of: a squeeze convolution layer (which has only 1×1 filters), feeding into an expand layer that has a mix of 1×1 and 3×3 convolution filters.
 - There are three tunable dimensions (hyperparameters) in a Fire module: s1×1, e1×1, and e3×3.
    - s1×1: The number of 1×1 in squeeze layer.
    - e1×1 and e3×3: The number of 1×1 and 3×3 in expand layer.
 - When we use Fire modules we set s1×1 to be less than (e1×1 + e3×3), so the squeeze layer helps to limit the number of input channels to the 3×3 filters, as per Strategy 2 in previous section.
 - To me, it is quite a like of [Inception](https://medium.com/coinmonks/paper-review-of-googlenet-inception-v1-winner-of-ilsvlc-2014-image-classification-c2b3565a64e7) Module.

**Comment by NorbertZheng** [comment]
## SqueezeNet Architecture
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2d8ef8cc-c809-4ff3-9f92-f57e596eb320)
SqueezeNet (Left), SqueezeNet with simple bypass (Middle), SqueezeNet with complex bypass (Right).

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/94859475-7b76-4dd2-9a35-61ea8d20157d)
Details of SqueezeNet Architecture.

 - SqueezeNet (Left): begins with a standalone convolution layer (conv1), followed by 8 Fire modules (fire2–9), ending with a final conv layer (conv10).
 - The number of filters per fire module is gradually increased from the beginning to the end of the network.
 - Max-pooling with a stride of 2 is performed after layers conv1, fire4, fire8, and conv10.
 - SqueezeNet with simple bypass (Middle) and SqueezeNet with complex bypass (Right): The use of bypass is inspired by [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8).

**Comment by NorbertZheng** [comment]
## Evaluation of SqueezeNet
### Comparing SqueezeNet to model compression approaches
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/2a6bd9bb-45c8-4bf5-855f-2dd888343efa)
Comparing SqueezeNet to model compression approaches.

 - **With SqueezeNet, we achieve a 50× reduction in model size compared to [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160), while meeting or exceeding the top-1 and top-5 accuracy of [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160).**
 - And the model size reduction is much higher than SVD, network pruning and deep compression.
 - Applying Deep Compression with 8-bit quantization, SqueezeNet yields a 0.66 MB model (363× smaller than 32-bit [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160)) with equivalent accuracy to [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160). Further, applying Deep Compression with 6-bit quantization and 33% sparsity on SqueezeNet, a 0.47MB model (510× smaller than 32-bit [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160)) with equivalent accuracy. SqueezeNet is indeed amenable to compression.

**Comment by NorbertZheng** [comment]
### Hyperparameters
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5ddd0ff1-5cc9-4963-998e-a1a7b92a05a5)
Different Hyperparameter Values for SqueezeNet.

 - Squeeze ratio (SR) (Left): the ratio between the number of filters in squeeze layers and the number of filters in expand layers.
 - Increasing SR beyond 0.125 can further increase ImageNet top-5 accuracy from 80.3% (i.e. [AlexNet](https://medium.com/coinmonks/paper-review-of-alexnet-caffenet-winner-in-ilsvrc-2012-image-classification-b93598314160)-level) with a 4.8MB model to 86.0% with a 19MB model. Accuracy plateaus at 86.0% with SR=0.75 (a 19MB model), and setting SR=1.0 further increases model size without improving accuracy.
 - Percentage of 3×3 Filters (Right): Top-5 accuracy plateaus at 85.6% using 50% 3×3 filters, and further increasing the percentage of 3×3 filters leads to a larger model size but provides no improvement in accuracy on ImageNet.

**Comment by NorbertZheng** [comment]
### SqueezeNet Variants
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/14e33e21-3ef3-4b94-bd69-df1568555b59)
SqueezeNet accuracy and model size using different macroarchitecture configurations.

 - Complex and simple bypass connections both yielded an accuracy improvement over the vanilla SqueezeNet architecture.
 - Interestingly, the simple bypass enabled a higher accuracy accuracy improvement than complex bypass.
 - Adding the simple bypass connections yielded an increase of 2.9 percentage-points in top-1 accuracy and 2.2 percentage-points in top-5 accuracy without increasing model size.

**Comment by NorbertZheng** [comment]
## Reference
 - [2016 arXiv] [SqueezeNet] [SqueezeNet: AlexNet-level accuracy with 50x fewer parameters and <0.5MB model size](https://arxiv.org/abs/1602.07360).
